ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Sync.hs

```diff
@@ -302,25 +321,109 @@ basicChainSyncClient tracer cfg chainDbView varCandidate =
     nullTracer
     varCandidate
 
+-- | Same as 'runConnectedPeersPipelined' except the client peer is ran not with
+-- 'runPipelinedPeer' but with 'runPipelinedPeerWithLimits'.
+--
+-- REVIEW: Should probably go to another file.
+runConnectedPeersPipelinedWithLimits ::
+  ( MonadAsync m
+  , MonadFork m
+  , MonadMask m
+  , MonadThrow (STM m)
+  , MonadTimer m
+  , Show failure
+  , forall (st' :: ps). Show (ClientHasAgency st')
+  , forall (st' :: ps). Show (ServerHasAgency st')
+  , ShowProxy ps
+  ) =>
+  m (Channel m bytes, Channel m bytes) ->
+  Tracer m (Role, TraceSendRecv ps) ->
+  TP.Codec ps failure m bytes ->
+  ProtocolSizeLimits ps bytes ->
+  ProtocolTimeLimits ps ->
+  TP.PeerPipelined ps pr st m a ->
+  TP.Peer ps (TP.FlipAgency pr) st m b ->
+  m (a, b)
+runConnectedPeersPipelinedWithLimits createChannels tracer codec sizeLimits timeLimits client server =
+    createChannels >>= \(clientChannel, serverChannel) ->
+    (fst <$> runPipelinedPeerWithLimits tracerClient codec sizeLimits timeLimits clientChannel client)
+      `concurrently`
+    (fst <$> runPeer tracerServer codec serverChannel server)
+  where
+    tracerClient = contramap ((,) Client) tracer
+    tracerServer = contramap ((,) Server) tracer
+
+chainSyncNoSizeLimits :: ProtocolSizeLimits (ChainSync header point tip) bytes
+chainSyncNoSizeLimits = byteLimitsChainSync (const 0)
+
+chainSyncTimeouts ::
+  SlotLength ->
+  Asc ->
+  ChainSyncTimeout
+chainSyncTimeouts t f =
+  ChainSyncTimeout{
+      canAwaitTimeout
+    , intersectTimeout
+    , mustReplyTimeout
+    }
+  where
+    canAwaitTimeout :: Maybe DiffTime
+    canAwaitTimeout = shortWait -- REVIEW: what is this exactly?
+
+    intersectTimeout :: Maybe DiffTime
+    intersectTimeout = shortWait -- REVIEW: what is this exactly?
+
+    -- | The following timeout is derived from the average length of a streak of
+    -- empty slots. If the probability of the election of a leader is @f@ and
+    -- @Y@ is a probability, then a streak of empty slots will be shorter than
+    -- @log (1 - Y) / log (1 - f)@ with probability @Y@. Main net nodes pick a
+    -- random value for @Y@ between 99.9% and 99.999%. For our use case, we
+    -- choose the tightest bound of 99.9%.
+    mustReplyTimeout :: Maybe DiffTime
+    mustReplyTimeout = Just $ fromIntegral $ round $
+      (realToFrac $ getSlotLength t)
+        * log (1 - 0.999) / log (1 - ascVal f)
```

Comment:
I get a warning in this function; this is what makes the CI unhappy. To be exact:

```
    • Defaulting the following constraint to type ‘Integer’
        Integral a0 arising from a use of ‘fromIntegral’
    • In the second argument of ‘($)’, namely
        ‘fromIntegral
           $ round
               $ (realToFrac $ getSlotLength t) * log (1 - 0.999)
                   / log (1 - ascVal f)’
      In the expression:
        Just
          $ fromIntegral
              $ round
                  $ (realToFrac $ getSlotLength t) * log (1 - 0.999)
                      / log (1 - ascVal f)
      In an equation for ‘mustReplyTimeout’:
          mustReplyTimeout
            = Just
                $ fromIntegral
                    $ round
                        $ (realToFrac $ getSlotLength t) * log (1 - 0.999)
                            / log (1 - ascVal f)
    |
383 |     mustReplyTimeout = Just $ fromIntegral $ round $
    |                               ^^^^^^^^^^^^^^^^^^^^^^...
```

I don't know how to handle this.

Reply:
Solved in https://github.com/input-output-hk/ouroboros-consensus/pull/371/commits/6fd61ff1f7fac0e5c7c7580b73beafde90016085.